### DISCUSSION TITLE
Why does strongswan reassemble the packets before adding ESP Header

### DISCUSSION THREAD

**Discussion by subramaniansairam** [strongswan/strongswan#1763] [Q&A]
In one of our testing of sending data from the SGW (Serving Gateway) towards the eNodeB through the StrongSwan, I see the following.

- From the FTP/SGW 2 fragments arrive at the Receiving side of the Security Gateway
- The Security Gateway, assembles these 2 packets as one, add the ESP Header and then sends it to the eNodeB
- Why is that the StrongSwan Security Gateway, reassembling the packets before adding the ESP Header, rather than adding the ESP Headers to both the packets and forward to the eNodeB?
- Is this behavior correct? If yes, then is there any way to avoid this using any configuration on the Security Gateway OR Linux server

**Comment by tobiasbrunner** [comment]
Please note that strongSwan generally doesn't handle IPsec traffic, this is handled by the Linux kernel.

Regardless of IPsec being used, I think the Linux kernel (or rather Netfilter) does this fragment reassembly for forwarded packets in pretty much all configurations (e.g. for stateful firewall rules or NATing) and just fragments them again if necessary/possible. It might be possible to avoid that somehow (e.g. with careful firewall rules), but not sure on the details.

For IPsec, there is a similar problem with fragments, as non-initial fragments can't fully be matched against IPsec policies as there is no transport protocol header (TCP/UDP) to match port selectors (you can read about that in [RFC 4301, section 7](https://datatracker.ietf.org/doc/html/rfc4301#section-7) and [Appendix D](https://datatracker.ietf.org/doc/html/rfc4301#page-88)). So forwarding non-initial fragments, if forwarded as such by the kernel, can only work if no port selectors are configured on the policies.

**Comment by Kentzo** [comment]
Could this be helpful: https://unix.stackexchange.com/questions/650790/unwanted-defragmentation-of-forwarded-ipv4-packets ?
